security/tpm/src/tpm.ts

```diff
@@ -3,22 +3,412 @@
+                    /*Codess_SRS_NODE_TPM_SECURITY_CLIENT_06_007: [Any errors from interacting with the TPM hardware will cause in SecurityDeviceError to be returned in the err parameter of the callback.] */
+                    this._fsm.transition('disconnected', callback, ekCreateErr);
+                  } else {
+                    /*Codes_SRS_NODE_TPM_SECURITY_CLIENT_06_006: [The `getEndorsmentKey` function shall query the TPM hardware and return the `endorsementKey` in the callback.] */
+                    this._ek = ekPublicKey;
+                    this._createPersistentPrimary('SRK', tss.Owner, TpmSecurityClient._srkPersistentHandle, TpmSecurityClient._srkTemplate, (srkCreateErr: Error, srkPublicKey: TPMT_PUBLIC) => {
+                      if (srkCreateErr) {
+                        /*Codes_SRS_NODE_TPM_SECURITY_CLIENT_06_009: [Any errors from interacting with the TPM hardware will cause in SecurityDeviceError to be returned in the err parameter of the callback.] */
+                        this._fsm.transition('disconnected', callback, srkCreateErr);
+                      } else {
+                        /*Codes_SRS_NODE_TPM_SECURITY_CLIENT_06_008: [The `getStorageRootKey` function shall query the TPM hardware and return the `storageRootKey` in the callback.] */
+                        this._srk = srkPublicKey;
+                        this._fsm.transition('connected', callback);
+                      }
+                    });
+                  }
+                });
+              });
+            } catch (err) {
+              this._fsm.transition('disconnected', callback, err);
+            }
+          },
+          '*': () => this._fsm.deferUntilTransition()
+        },
+        connected: {
+          _onEnter: (callback) => {
+            callback(null);
+          },
+          getEndorsementKey: (callback) => {
+            callback(null, this._ek.asTpm2B());
+          },
+          getStorageRootKey: (callback) => {
+            callback(null, this._srk.asTpm2B());
+          },
+          signWithIdentity: (dataToSign, callback) => {
+            this._signData(dataToSign, (err: Error, signedData: Buffer) => {
+              if (err) {
+                debug('Error from signing data: ' + err);
+                this._fsm.transition('disconnected', callback, err);
+              } else {
+                callback(null, signedData);
+              }
+            });
+          },
+          activateSymmetricIdentity: (identityKey, callback) => {
+            this._activateSymmetricIdentity(identityKey, (err: Error) => {
+              if (err) {
+                debug('Error from activate: ' + err);
+                this._fsm.transition('disconnected', callback, err);
+              } else {
+                callback(null);
+              }
+            });
+          },
+        }
+      }
+    });
+  }
+
+  getEndorsementKey(callback: (err: Error, endorsementKey: Buffer) => void): void {
```

Comment:
>endorsementKey [](start = 43, length = 14)

guessing those second arguments in callbacks should be optional in case there's an error

Reply:
Done.

---
In reply to: [160287911](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r160287911) [](ancestors = 160287911)